core/src/main/scala/scalafix/rewrite/Xor2Either.scala

```diff
@@ -1,28 +1,55 @@
 package scalafix.rewrite
 
+import scala.collection.immutable.{ListMap, Seq}
 import scala.meta.parsers.Parse
 import scala.{meta => m}
-import scalafix.util.Patch
+import scalafix.util._
+import PartialFunction._
+import scala.meta.Type.Select
+import scala.meta._
+
+class Desugared[T <: Tree: Parse](implicit semantic: SemanticApi) {
+  def unapply(original: T): Option[T] = semantic.desugared(original)
+}
 
 case object Xor2Either extends Rewrite {
   override def rewrite(ast: m.Tree, ctx: RewriteCtx): Seq[Patch] = {
-    import scala.meta._
-    val semantic = getSemanticApi(ctx)
-    class Desugared[T <: Tree: Parse] {
-      def unapply(original: T): Option[T] = semantic.desugared(original)
-    }
+    implicit val semanticApi: SemanticApi = getSemanticApi(ctx)
     object DType extends Desugared[Type]
     object DTerm extends Desugared[Term]
-    // NOTE. This approach is super inefficient, since we run semantic.desugar on
-    // every case for every node in the tree. Ideally, we first match on the
-    // syntax structure we want and then run semantic.desugar.
-    ast.collect {
-      case t @ DType(t"cats.data.Xor") =>
-        Patch(t.tokens.head, t.tokens.last, s"Either")
-      case t @ DTerm(q"cats.data.Xor.Right.apply[..$_]") =>
-        Patch(t.tokens.head, t.tokens.last, s"Right")
-      case t @ DTerm(q"cats.data.Xor.Left.apply[..$_]") =>
-        Patch(t.tokens.head, t.tokens.last, s"Left")
-    }
+
+    val typeChanger = new ChangeType(ast, ctx)
+    val methodChanger = new ChangeMethod(ast, ctx)
+    val importAdder = new AddImport(ast, ctx)
+
+    //Create a sequence of type replacements
+    val replacementTypes = List(
+      ReplaceType(t"cats.data.XorT", t"cats.data.EitherT", "EitherT"),
+      ReplaceType(t"cats.data.Xor", t"scala.util.Either", "Either"),
+      ReplaceType(t"cats.data.Xor.Left", t"scala.util.Left", "Left"),
+      ReplaceType(t"cats.data.Xor.Right", t"scala.util.Either.Right", "Right")
+    )
```

Comment:
The biggest issue being, now the heck to I make them compose?  Say I used some sort of path to represent "scala.util.Right.apply".  With /scala/util/Right/apply it's trivial to get the name, the fully qualified name, the parent, the parents name alone, or full path, etc.  And if I make a function to get "name" which just does path.takeRight(1) and another to get parent which is path.init, it's easy to compose them and get parent's name.  With partial functions, quasiquotes, and pattern matching I am finding the few bits of scala syntax I have run into that really resist composition.  Pointers hugely appreciated.